bionic/deriver.py

```diff
@@ -601,8 +591,74 @@ def __init__(self, task):
 
         self.is_complete = False
 
+    def incomplete_dep_states(self):
+        return [dep_state for dep_state in self.dep_states if not dep_state.is_complete]
+
+    @property
     def is_blocked(self):
-        return not all(parent.is_complete for parent in self.parents)
+        return len(self.incomplete_dep_states()) > 0
 
     def __repr__(self):
         return f"TaskState({self.task!r})"
+
+
+class Blockage:
+    """
+    Represents a blocking relationship between a task state and a collection of
+    not-yet-completed task keys it depends on.
+    """
+
+    def __init__(self, blocked_state, blocking_tks):
+        self.blocked_state = blocked_state
+        self._blocking_tks = set(blocking_tks)
+
+    def mark_task_key_complete(self, blocking_tk):
+        self._blocking_tks.discard(blocking_tk)
+
+    def is_resolved(self):
+        return not self._blocking_tks
+
+
+class BlockageTracker:
+    """
+    A helper class that keeps track of which task states are blocked by others.
+
+    A task state X is "blocked" by another task state Y if X depends on Y and Y is
+    not complete.
+    """
+
+    def __init__(self):
+        self._blockage_lists_by_blocking_tk = defaultdict(list)
+
+    def add_blockage(self, blocked_state, blocking_states):
+        """Records the fact that one task state is blocked by certain others."""
+
+        blocking_tks = [
+            blocking_tk
+            for blocking_state in blocking_states
+            for blocking_tk in blocking_state.task.keys
+        ]
+        blockage = Blockage(blocked_state, blocking_tks)
+        for blocking_tk in blocking_tks:
+            self._blockage_lists_by_blocking_tk[blocking_tk].append(blockage)
+
+    def get_unblocked_by(self, completed_state):
+        """
+        Records the fact that a task state is complete, and yields all task states
+        that are newly unblocked.
+        """
+
+        for completed_tk in completed_state.task.keys:
+            affected_blockages = self._blockage_lists_by_blocking_tk[completed_tk]
+            for blockage in affected_blockages:
+                blockage.mark_task_key_complete(completed_tk)
+                if blockage.is_resolved():
+                    yield blockage.blocked_state
+
+    def get_all_blocked_states(self):
+        return [
```

Comment:
I think this would duplicates since a blockage can be in multiple task keys. How do you feel about returning a set instead?

Reply:
Sure, that's reasonable. (In practice I doubt it will matter but it makes sense to avoid it.)